### ISSUE TITLE
Wrong "lifetime may not live long enough" diagnostic

### ISSUE THREAD

**Issue by paulocsanz** [rust-lang/rust#99303]
The compiler is misdiagnosing some compilation errors as lifetime errors. This has happened 4 times today, but it had happened before in the past usually once a week or less.

A cargo clean + check reports a different error (which was the actual problem). Sadly we don't know how to minimally reproduce it, the things we found were: it always wrongly points to a lifetime problem during the creation of a `Context<'_, '_>` from another one, so we assume lifetime deduction is at fault. And it always points to places that we haven't touched in a while.

We have been seeing this happen more after we started to heavily use `Box<dyn Trait + Send + Sync>`, so it might also be related to that, but we don't have enough evidence to be sure.

And we use cargo workspaces heavily. Our codebase is very complex and big, but it's private so I can't share it, we do trigger a lot of incremental compilation problems that generally cause a crash and require a cargo clean, but this is a new one.

The wrong error message:

```
error: lifetime may not live long enough
   --> lib/dal/src/context.rs:273:56
    |
270 |           &self,
    |           - let's call the lifetime of this reference `'4`
271 |           ctx: &DalContext<'_, '_>,
    |                            -- let's call the lifetime of this reference `'2`
272 |           oid: OrganizationId,
273 |       ) -> Result<DalContext<'_, '_>, TransactionsError> {
    |  ________________________________________________________^
274 | |         let mut new = self.clone();
275 | |         new.update_to_organization_tenancies(ctx, oid).await?;
276 | |         Ok(new)
277 | |     }
    | |_____^ associated function was supposed to return data with lifetime `'2` but it is returning data with lifetime `'4`

error: lifetime may not live long enough
   --> lib/dal/src/context.rs:273:56
    |
270 |           &self,
    |           - let's call the lifetime of this reference `'4`
271 |           ctx: &DalContext<'_, '_>,
    |                            -- let's call the lifetime of this reference `'2`
272 |           oid: OrganizationId,
273 |       ) -> Result<DalContext<'_, '_>, TransactionsError> {
    |  ________________________________________________________^
274 | |         let mut new = self.clone();
275 | |         new.update_to_organization_tenancies(ctx, oid).await?;
276 | |         Ok(new)
277 | |     }
    | |_____^ associated function was supposed to return data with lifetime `'4` but it is returning data with lifetime `'2`

error: lifetime may not live long enough
   --> lib/dal/src/context.rs:276:9
    |
270 |         &self,
    |         ----- has type `&DalContext<'1, '_>`
271 |         ctx: &DalContext<'_, '_>,
    |                          -- let's call the lifetime of this reference `'2`
...
276 |         Ok(new)
    |         ^^^^^^^ associated function was supposed to return data with lifetime `'2` but it is returning data with lifetime `'1`

error: lifetime may not live long enough
   --> lib/dal/src/context.rs:276:9
    |
270 |         &self,
    |         ----- has type `&DalContext<'_, '3>`
271 |         ctx: &DalContext<'_, '_>,
    |                          -- let's call the lifetime of this reference `'2`
...
276 |         Ok(new)
    |         ^^^^^^^ associated function was supposed to return data with lifetime `'2` but it is returning data with lifetime `'3`

warning: `dal` (lib) generated 4 warnings
error: could not compile `dal` due to 4 previous errors; 4 warnings emitted
```

The error message after `cargo clean` + `cargo check`:

```
error[E0507]: cannot move out of `self.access_builder` which is behind a shared reference
  --> lib/dal/src/job/definition/component_post_processing.rs:77:34
   |
77 |             serde_json::to_value(self.access_builder)?,
   |                                  ^^^^^^^^^^^^^^^^^^^ move occurs because `self.access_builder` has type `AccessBuilder`, which does not implement the `Copy` trait
```

### Meta

`rustc --version --verbose`:
```
rustc 1.62.0 (a8314ef7d 2022-06-27)
binary: rustc
commit-hash: a8314ef7d0ec7b75c336af2c9857bfaf43002bfc
commit-date: 2022-06-27
host: aarch64-apple-darwin
release: 1.62.0
LLVM version: 14.0.5
```

**Comment by jhelwig** [comment]
Another variant of the same kind of behavior (lifetime error that goes away with a `cargo clean`) that we've hit in the same codebase @paulocsanz mentioned in the initial report:

```
error[E0424]: expected value, found module `self`
   --> lib/sdf/src/server/server.rs:440:45
    |
428 | fn faktory_job_wrapper(
    |    ------------------- this function can't have a `self` parameter
...
440 |                 let ctx = ctx_builder.build(self.access_builder().build(self.visibility()), &txns);
    |                                             ^^^^ `self` value is a keyword only available in methods with a `self` parameter

error[E0424]: expected value, found module `self`
   --> lib/sdf/src/server/server.rs:440:73
    |
428 | fn faktory_job_wrapper(
    |    ------------------- this function can't have a `self` parameter
...
440 |                 let ctx = ctx_builder.build(self.access_builder().build(self.visibility()), &txns);
    |                                                                         ^^^^ `self` value is a keyword only available in methods with a `self` parameter

For more information about this error, try `rustc --explain E0424`.
```

After the `cargo clean` with this variant, there were no errors in the subsequent `cargo check`.

**Comment by paulocsanz** [comment]
Closing this we haven't seen that anymore and 1.62.1 has a changelog that seems to fix a similar issue, so that might be it. Will reopen if it happens again.
